### PR TITLE
Added EAP_ANYCONNECT support

### PR DESCRIPTION
**Pull request by steweg**

### CODE REVIEW

**Comment by Thermi** [comment]
Please make separate commits by functionality and rebase onto latest master.
The if_id support for ipsec.conf is unlikely to be merged because stroke (and hence ipsec.conf) are gone in the next major version.

File: src/libcharon/Makefile.am [hunk 1]
```diff
@@ -411,6 +411,20 @@ if MONOLITHIC
...
 endif
 
+if USE_EAP_ANYCONNECT
+  SUBDIRS += plugins/eap_anyconnect
+if MONOLITHIC
+  libcharon_la_LIBADD += plugins/eap_anyconnect/libstrongswan-eap-anyconnect.la
+endif
+endif
+
+if USE_EAP_MD5
```
Comment (Thermi) [inline src/libcharon/Makefile.am:421]: That's duplicated from above.

File: man/ipsec.conf.5.in [hunk 2]
```diff
@@ -471,6 +471,16 @@ Note that fragmented IKE messages sent by a peer are always accepted
 irrespective of the value of this option (even when set to
 .BR no ).
 .TP
+.BR if_id_in " = <id>"
```
Comment (Thermi) [inline man/ipsec.conf.5.in:474]: The whole ipsec.conf support for if_id is probably not going to be merged.

File: src/libcharon/encoding/payloads/eap_payload.c [hunk 3]
```diff
@@ -245,11 +246,11 @@ static size_t extract_type(private_eap_payload_t *this, size_t offset,
 }
 
 METHOD(eap_payload_t, get_type, eap_type_t,
-	private_eap_payload_t *this, uint32_t *vendor)
```
Comment (Thermi) [inline src/libcharon/encoding/payloads/eap_payload.c:248]: Can you separate out all the `uint32_t` to `eap_vendor_t` rewrites you did in its own commit please?

File: src/libcharon/plugins/stroke/stroke_config.c [hunk 4]
```diff
@@ -418,14 +418,30 @@ static auth_cfg_t *build_auth_cfg(private_stroke_config_t *this,
...
 		certificate_t *first = NULL;
+		certificate_t *last = NULL;
 
 		enumerator = enumerator_create_token(cert, ",", " ");
 		while (enumerator->enumerate(enumerator, &cert))
 		{
 			certificate = this->cred->load_peer(this->cred, cert);
 			if (certificate)
 			{
-				cfg->add(cfg, AUTH_RULE_SUBJECT_CERT, certificate);
```
Comment (Thermi) [inline src/libcharon/plugins/stroke/stroke_config.c:428]: What exactly is this used for?

File: src/libcharon/sa/eap/eap_method.h [hunk 5]
```diff
@@ -152,6 +152,27 @@ struct eap_method_t {
...
+
+	/**
+	 * Sets nonce used for subsequent calculation of signatures
+	 *
+	 * @param nonce			The nonce
+	 */
+	void (*set_nonce) (eap_method_t *this, chunk_t nonce);
+
+	/**
+	 * Sets IKE_SA_INIT message for subsequnet calculation of sugnatures
```
Comment (Thermi) [inline src/libcharon/sa/eap/eap_method.h:164]: Typos:
`sugnatures` -> `signatures`
`subsequnet` -> `subsequent`

File: src/libcharon/sa/ikev2/authenticators/pubkey_authenticator.c [hunk 6]
```diff
@@ -439,9 +439,31 @@ static status_t sign_classic(private_pubkey_authenticator_t *this,
 	switch (private->get_type(private))
 	{
 		case KEY_RSA:
-			scheme = SIGN_RSA_EMSA_PKCS1_SHA1;
+		{
```
Comment (Thermi) [inline src/libcharon/sa/ikev2/authenticators/pubkey_authenticator.c:442]: Please break this out into its own commit.

File: src/libcharon/sa/ikev2/tasks/ike_vendor.c [hunk 7]
```diff
@@ -145,12 +153,41 @@ static vid_data_t vids[] = {
...
+	if (cisco_anyconnect && message->get_exchange_type(message) == IKE_AUTH)
+	{
+		if (message->get_message_id(message) != 1)
+		{
+			DBG2(DBG_IKE, "skipping building of vendor IDs due to ANYCONNECT processing");
+			return this->initiator ? NEED_MORE : SUCCESS;
+		}
+		else
+		{
+			DBG2(DBG_IKE, "generating vendor ID for ANYCONNECT emulation");
```
Comment (Thermi) [inline src/libcharon/sa/ikev2/tasks/ike_vendor.c:175]: Why exactly the random vendor id?

File: src/libcharon/sa/ikev2/tasks/ike_vendor.c [hunk 8]
```diff
@@ -97,14 +97,22 @@ static vid_data_t vids[] = {
 	/* strongSwan MD5("strongSwan") */
 	{ "strongSwan", EXT_STRONGSWAN, "send_vendor_id", FALSE, 16,
 	  "\x88\x2f\xe5\x6d\x6f\xd2\x0d\xbc\x22\x51\x61\x3b\x2e\xbe\x5b\xeb"},
-	{ "Cisco Delete Reason", 0, NULL, FALSE, 0,
+	{ "Cisco Delete Reason", 0, "cisco_anyconnect", FALSE, 0,
```
Comment (Thermi) [inline src/libcharon/sa/ikev2/tasks/ike_vendor.c:100]: Please separate out the new vendor IDs into one commit.

**Review by Thermi — COMMENTED** [review]
I added some comments regarding some issues I found and open questions.
Do you have some source for the "questionable" stuff like the usage of IKEv1 fragmentation format in IKEv2, and the usage of a random vendor id?

Also: This is just stuff I found. If the code is merged is not exactly up to me, but it is extremely likely that these are issues that are going to be commented on by Tobias, too.

File: src/libcharon/encoding/payloads/eap_payload.c (hunk 3 above)
Comment (steweg) [reply src/libcharon/encoding/payloads/eap_payload.c:248]: Ok

File: src/libcharon/plugins/stroke/stroke_config.c (hunk 4 above)
Comment (steweg) [reply src/libcharon/plugins/stroke/stroke_config.c:428]: eap-anyconnect at certain point expect to send whole trustchain. Unfortunately the IM certificates are not loaded anywhere, therefore I have added this option to stroke (and vici). I will put it in separate commit

File: src/libcharon/sa/eap/eap_method.h (hunk 5 above)
Comment (steweg) [reply src/libcharon/sa/eap/eap_method.h]: Fixed

File: man/ipsec.conf.5.in (hunk 2 above)
Comment (steweg) [reply man/ipsec.conf.5.in:474]: Ok

File: src/libcharon/Makefile.am (hunk 1 above)
Comment (steweg) [reply src/libcharon/Makefile.am]: Fixed

**Comment by steweg** [comment]
As suggested I will separate those commit into separate pull requests
